File: src/plugins/slash_commands.cpp
#include "slash_commands.hpp"


bool SlashCommandsPlugin::event_Chat(const ClientMsg_Chat &msg) {
    std::vector<std::string> tokens = PluginAPI::tokenizeMessage(std::string(msg.text));

    if (tokens.size() == 0) {
        return false;
    }

    const std::string &command = tokens[0];

    if (command[0] != '/') {
        return false;
    }

    if (command == "/set") {
        if (tokens.size() < 3) {
            PluginAPI::sendChatMessage(SERVER_CLIENT_ID, msg.clientId, "Usage: /set <setting> <value>");
            return true;
        }

        std::string key = tokens[1];
        float value = std::stof(tokens[2]);

        PluginAPI::setWorldSetting(key, value);
        PluginAPI::sendChatMessage(SERVER_CLIENT_ID, BROADCAST_CLIENT_ID, "Setting " + key + " set to " + std::to_string(value));
    } else if (command == "/kill") {
        if (tokens.size() < 2) {
            PluginAPI::sendChatMessage(SERVER_CLIENT_ID, msg.clientId, "Usage: /kill <playername>");
            return true;
        }

        std::string targetName = tokens[1];
        client_id targetId = PluginAPI::getPlayerByName(targetName);
        if (targetId != 0) {
            PluginAPI::killPlayer(targetId);
        }
    } else if (command == "/playerlist") {
        PluginAPI::sendChatMessage(SERVER_CLIENT_ID, msg.clientId, "Online Players:");
        std::vector<client_id> playerIds = PluginAPI::getAllPlayerIds();
        for (client_id pid : playerIds) {
            std::string playerName = PluginAPI::getPlayerName(pid);
            std::string line = "- " + playerName + " (ID: " + std::to_string(pid) + ")";
            line += " @ " + PluginAPI::getPlayerIP(pid);
            PluginAPI::sendChatMessage(SERVER_CLIENT_ID, msg.clientId, line);
        }
    } else {
        PluginAPI::sendChatMessage(SERVER_CLIENT_ID, msg.clientId, "Unknown command: " + command);
    }

    return true;
};


REGISTER_PLUGIN(SlashCommandsPlugin)

File: plugins/slash_commands.py
import bzapi

def slash_command_handler(from_id: int, to_id: int, text: str) -> None:
    tokens = text.split()

    if tokens[0] == '/setPlayerParameter':
        if len(tokens) != 4:
            bzapi.send_chat_message(0, from_id, "Usage: /setPlayerParameter <client_id> <parameter> <value>")
            return True

        target_id = int(tokens[1])
        parameter = tokens[2]
        try:
            value = float(tokens[3])
        except ValueError:
            bzapi.send_chat_message(0, from_id, "Value must be a number.")
            return True

        if not bzapi.set_player_parameter(target_id, parameter, value):
            bzapi.send_chat_message(0, from_id, f"Failed to set parameter '{parameter}' for player ID {target_id}.")
        else:
            bzapi.send_chat_message(0, from_id, f"Player parameter '{parameter}' set to {value}.")
        return True
    
    elif tokens[0] == '/kill':
        if len(tokens) != 2:
            bzapi.send_chat_message(0, from_id, "Usage: /kill <player_name>")
            return True

        target_name = tokens[1]
        target_id = bzapi.get_player_by_name(target_name)


        if target_id is None:
            bzapi.send_chat_message(0, from_id, f"Player '{target_name}' not found.")
            return True

        bzapi.kill_player(target_id)
        bzapi.send_chat_message(0, from_id, f"Player '{target_name}' has been killed.")
        return True

    elif tokens[0] == '/listPlayers':
        player_list = bzapi.get_all_player_ids()
        
        for pid in player_list:
            pname = bzapi.get_player_name(pid)
            pip = bzapi.get_player_ip(pid)
            bzapi.send_chat_message(0, from_id, f"Player ID: {pid}, Name: {pname}, IP: {pip}")

        return True
    
    elif tokens[0][0] == '/':
        bzapi.send_chat_message(0, from_id, f"Unknown command: {tokens[0]}")
        return True
    
    return False

bzapi.register_callback(bzapi.event_type.CHAT, slash_command_handler)
